controllers/oscmachine_vm_controller.go

```diff
@@ -602,6 +603,7 @@ func reconcileDeleteVm(ctx context.Context, clusterScope *scope.ClusterScope, ma
 	}
 	if vmId == "" {
 		log.V(2).Info("vm is already destroyed", "vmName", vmName)
+		controllerutil.RemoveFinalizer(machineScope.OscMachine, "")
```

Comment:
This removes nothing, the finalizer name is required:

```
controllerutil.RemoveFinalizer(machineScope.OscMachine, "oscmachine.infrastructure.cluster.x-k8s.io")
```

Even if it did something, I don't see why you need to add this. The caller function removes the finalizer (unless there is a problem with the keypair deletion).